### PR TITLE
OCPBUGS-33129: Panic when we remove an OCB infra MCP and we try to create new ones with different names

### PR DESCRIPTION
**Pull request by inesqyx**
Avoid non-existing MOSC de-reference.

### CODE REVIEW

File: pkg/controller/node/node_controller.go [hunk 1]
```diff
@@ -381,38 +382,60 @@ func (ctrl *Controller) makeMasterNodeSchedulable(node *corev1.Node) error {
...
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get MachineOSConfig from MachineOSBuild %#v", curMOSB))
+		return
+	}
+	if config == nil {
+		ctrl.client.MachineconfigurationV1alpha1().MachineOSBuilds().Delete(context.TODO(), curMOSB.Name, metav1.DeleteOptions{})
+		klog.V(4).Infof("Deleting MachineOSBuild %s due to missing MOSC", curMOSB.Name)
+		return
+	}
 	mcp, _ := ctrl.mcpLister.Get(config.Spec.MachineConfigPool.Name)
```
[the remark this answers is not in the document]
Comment (inesqyx) [reply pkg/controller/node/node_controller.go:397]: I think the panic is mainly from dereferencing an empty MOSC to find the MCP. So an empty MCP is fine while a MOSC is not. We should have a mechanism to delete MOSB when the MOSC is deleted but it seems there are corner cases here. So I have added that whenever we check for a MOSB, we check its MOSC first and if that is not found, the MOSB should not exist either.

File: pkg/controller/node/node_controller.go [hunk 2]
```diff
@@ -381,38 +382,60 @@ func (ctrl *Controller) makeMasterNodeSchedulable(node *corev1.Node) error {
...
+	klog.V(4).Infof("Adding MachineOSBuild %s", curMOSB.Name)
 
-	config, _ := ctrl.client.MachineconfigurationV1alpha1().MachineOSConfigs().Get(context.TODO(), curMOSB.Spec.MachineOSConfig.Name, metav1.GetOptions{})
-
+	config, err := ctrl.client.MachineconfigurationV1alpha1().MachineOSConfigs().Get(context.TODO(), curMOSB.Spec.MachineOSConfig.Name, metav1.GetOptions{})
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get MachineOSConfig from MachineOSBuild %#v", curMOSB))
+		return
+	}
+	if config == nil {
```
Comment (deads2k) [inline pkg/controller/node/node_controller.go:392]: won't be nil unless there was an error.  Is this delete important?  

keep in mind that there is no synchronization between two different caches made by different list/watches.  This is a risky move. 

Can you separate the delete from the panic fix?

**Comment by sergiordlr** [comment]
I've tested the previous commit  8ecaa8d . The panic error is not in the machine-config-controller pod anymore, but we can find a panic in the machine-os-builder pod.

This is the panic logs


```
E0610 14:17:43.884756       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 92 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1fd8000?, 0x3962a90})
        /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000489ea0?})
        /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x1fd8000?, 0x3962a90?})
        /usr/lib/golang/src/runtime/panic.go:914 +0x21f
github.com/openshift/machine-config-operator/pkg/controller/common.(*MachineOSBuildState).IsBuilding(...)
        /go/src/github.com/openshift/machine-config-operator/pkg/controller/common/mos_state.go:71
github.com/openshift/machine-config-operator/pkg/controller/build.(*Controller).customBuildPodUpdater(0xc0007adea0, 0xc0008c4000)
        /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/build_controller.go:430 +0x4b2
github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).syncPod(0xc00037fe30, {0xc000591980, 0x57})
        /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:159 +0x610
github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).processNextWorkItem(0xc00037fe30)
        /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:332 +0xc4
github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).worker(...)
        /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:321
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
        /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x27024a0, 0xc0007811d0}, 0x1, 0xc000165380)
        /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
        /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0x0?, 0x0?)
        /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x1e
created by github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).Run in goroutine 45
        /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:183 +0x1e5
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x148 pc=0x1ceda32]

```

**Comment by sinnykumari** [comment]
LGTM
pending QE verification

**Comment by inesqyx** [comment]
> I've tested the previous commit [8ecaa8d](https://github.com/openshift/machine-config-operator/commit/8ecaa8d1bcaf2cff9e25e9e7da4287faa5b538e0) . The panic error is not in the machine-config-controller pod anymore, but we can find a panic in the machine-os-builder pod.
> 
> This is the panic logs
> 
> ```
> E0610 14:17:43.884756       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
> goroutine 92 [running]:
> k8s.io/apimachinery/pkg/util/runtime.logPanic({0x1fd8000?, 0x3962a90})
>         /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x85
> k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc000489ea0?})
>         /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x6b
> panic({0x1fd8000?, 0x3962a90?})
>         /usr/lib/golang/src/runtime/panic.go:914 +0x21f
> github.com/openshift/machine-config-operator/pkg/controller/common.(*MachineOSBuildState).IsBuilding(...)
>         /go/src/github.com/openshift/machine-config-operator/pkg/controller/common/mos_state.go:71
> github.com/openshift/machine-config-operator/pkg/controller/build.(*Controller).customBuildPodUpdater(0xc0007adea0, 0xc0008c4000)
>         /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/build_controller.go:430 +0x4b2
> github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).syncPod(0xc00037fe30, {0xc000591980, 0x57})
>         /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:159 +0x610
> github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).processNextWorkItem(0xc00037fe30)
>         /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:332 +0xc4
> github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).worker(...)
>         /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:321
> k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
>         /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
> k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x0?, {0x27024a0, 0xc0007811d0}, 0x1, 0xc000165380)
>         /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
> k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x0?, 0x3b9aca00, 0x0, 0x0?, 0x0?)
>         /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
> k8s.io/apimachinery/pkg/util/wait.Until(0x0?, 0x0?, 0x0?)
>         /go/src/github.com/openshift/machine-config-operator/vendor/k8s.io/apimachinery/pkg/util/wait/backoff.go:161 +0x1e
> created by github.com/openshift/machine-config-operator/pkg/controller/build.(*PodBuildController).Run in goroutine 45
>         /go/src/github.com/openshift/machine-config-operator/pkg/controller/build/pod_build_controller.go:183 +0x1e5
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
>         panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x148 pc=0x1ceda32]
> ```
> 
> I was able to reproduce it with commit [0311f03](https://github.com/openshift/machine-config-operator/commit/0311f03d7460da3aff7b48a99760eea45f308124) .
> 
> It seems that we can only reproduce itermittently though. It took me 2 tries to reproduce the panic.
> 
> Same steps
> 
> 1. Create infra pool
> 2. Create MOSC for infra pool
> 3. Wait for build pod to run
> 4. Remove MSOC
> 5. A panic happens in machine-os-builder pod (instead of MCC).

I think I figured out where went wrong here. So two panic cases comes from:
(1) MOSC got deleted during build, de-reference the MOSC to find the MOSB for updateMachineOSBuild panics 
(2) MOSC got deleted during build and MOSB got garbage collected, de-reference the MOSB to read build status panics 
The reason behind why MOSC is deleted in the middle if not manual and intentional needs further investigation.

**Comment by inesqyx** [comment]
Current behaviour:

-  Create infra pool
-  Create MOSC for infra pool
-  Wait for build pod to run
- Remove MOSC
- MOSB got deleted - garbage collection 
- machine-os-builder shut down 
- build pod shut down 
- no panic observed

**Comment by sergiordlr** [comment]
Verified using IPI on AWS


1. Create an infra pool
2. Create a MOSC resource pointing to the infra pool
3. Wait for the build pod to be running
4. Remove the the MOSC resource

No panic happened in the controller pod or the machine-os-builder pod.

We can see this log in the machine-os-builder pod instead of the panic


`I0611 10:21:16.101690       1 pod_build_controller.go:296] Error syncing pod openshift-machine-config-operator/build-rendered-infra-ad00d24655a1d28893279f65a6b807f1: unable to update with build pod status: Missing MOSC/MOSB for pool infra`


/label qe-approved

**Comment by sinnykumari** [comment]
Thanks Ines for working on the fix and Sergio for verifying the fix!
/lgtm

**Pull request merged**